core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/GraphStructure.java

```diff
@@ -99,12 +128,33 @@ public static <Solution_> GraphStructureAndDirection determineGraphStructure(
             }
         }
 
+        if (isArbitrary) {
+            return arbitraryGraphStructure;
+        }
+
         if (directionalType == null) {
             return new GraphStructureAndDirection(NO_DYNAMIC_EDGES, null, null);
         } else {
             // Cannot use a single successor function if there are multiple entity classes
-            return multipleDeclarativeEntityClasses ? new GraphStructureAndDirection(ARBITRARY, null, null)
-                    : new GraphStructureAndDirection(SINGLE_DIRECTIONAL_PARENT, parentMetaModel, directionalType);
+            return new GraphStructureAndDirection(SINGLE_DIRECTIONAL_PARENT, parentMetaModel, directionalType);
+        }
+    }
+
+    private static <Solution_> boolean doesEntitiesUseDeclarativeShadowVariables(Object[] entities,
```

Comment:
`doEntities...`?
Also, maybe switch the order of the arguments, and make entities a vararg?